rust/src/passwd.rs

```diff
@@ -173,16 +174,152 @@ pub fn migrate_group_except_root(rootfs_dfd: i32, preserved_groups: &Vec<String>
+/// cases we inject data into the tree before even laying
+/// down any files, and notably before running RPM `useradd` etc.
+pub fn passwd_compose_prep_repo(
     rootfs_dfd: i32,
+    treefile: &mut Treefile,
     mut ffi_repo: Pin<&mut crate::ffi::OstreeRepo>,
     previous_checksum: &str,
+    unified_core: bool,
 ) -> Result<()> {
-    anyhow::ensure!(!previous_checksum.is_empty(), "missing previous reference");
-
     let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
     let repo = ffi_repo.gobj_wrap();
+    passwd_compose_prep_impl(
+        &rootfs,
+        treefile,
+        Some((&repo, previous_checksum)),
+        unified_core,
+    )
+}
+
+fn passwd_compose_prep_impl(
+    rootfs: &openat::Dir,
+    treefile: &mut Treefile,
+    repo_previous_ref: Option<(&ostree::Repo, &str)>,
+    unified_core: bool,
+) -> Result<()> {
+    let generate_from_previous = treefile.parsed.preserve_passwd.unwrap_or(true);
+    if !generate_from_previous {
+        // Nothing to do
+        return Ok(());
+    };
+
+    let dest = if unified_core { "usr/etc/" } else { "etc/" };
+
+    // Create /etc in the target root; FIXME - should ensure we're using
+    // the right permissions from the filesystem RPM.  Doing this right
+    // is really hard because filesystem depends on setup which installs
+    // the files...
+    rootfs.ensure_dir_all(dest, 0o0755)?;
+
+    let found_passwd_data = data_from_json(rootfs, treefile, dest, "passwd")?;
+    let found_groups_data = data_from_json(rootfs, treefile, dest, "group")?;
+
+    // We should error if we are getting passwd data from JSON and group from
+    // previous commit, or vice versa, as that'll confuse everyone when it goes
+    // wrong.
+    match (found_passwd_data, found_groups_data) {
+        (true, false) => {
+            anyhow::bail!("configured to migrate passwd data from JSON, and group data from commit")
+        }
+        (false, true) => {
+            anyhow::bail!("configured to migrate passwd data from commit, and group data from JSON")
+        }
+        _ => {}
+    };
+
+    if !found_passwd_data {
+        if let Some((repo, prev_ref)) = repo_previous_ref {
+            concat_fs_content(&rootfs, repo, prev_ref)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn data_from_json(
+    rootfs: &openat::Dir,
+    treefile: &mut Treefile,
+    dest_path: &str,
+    target: &str,
+) -> Result<bool> {
+    anyhow::ensure!(!dest_path.is_empty(), "missing destination path");
+
+    let append_unique_entries = match target {
+        "passwd" => passwd_append_unique,
+        "group" => group_append_unique,
+        x => anyhow::bail!("invalid merge target '{}'", x),
+    };
+
+    let target_etc_filename = format!("{}{}", dest_path, target);
+
+    // Migrate the check data from the specified file to /etc.
+    let src_fd = if target == "passwd" {
+        let check_passwd = match treefile.parsed.check_passwd {
+            None => return Ok(false),
```

Comment:
I wonder if `Result<Option<()>>` might be more descriptive here?  Since `Some(())` more clearly means "we did/found something", whereas `bool` can easily be interpreted as either "we did/found something" or "we *didn't* do/find something".

(This is an...optional...comment)

Reply:
Possibly yes, I'll have a look.

Overall this logic is pretty unidiomatic and close to C, plus I'm learning the treefile field semantics as I go through it, I totally agree the result may look a bit weird.

I feel there is some room to unify and encapsulate this into something like an `Option<(passwd_content, group_content)>` directly in the treefile, but I didn't want to introduce to much novelty into here.
We may consider a final last pass to make the whole file less C-like, when all the logic has been moved.

Reply:
Oh yep, translating from C to first is totally fine!